File: v1/src/main/java/com/api/v1/customers/services/CustomerRegistrationServiceImpl.java
package com.api.v1.customers.services;

import com.api.v1.customers.Customer;
import com.api.v1.customers.domain.CustomerCrudRepository;
import com.api.v1.customers.CustomerResponseDTO;
import com.api.v1.people.Person;
import com.api.v1.people.exceptions.DuplicatedEmailException;
import com.api.v1.people.exceptions.DuplicatedSINException;
import com.api.v1.people.PersonRegistrationDTO;
import com.api.v1.people.PersonRegistrationService;
import jakarta.validation.Valid;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.net.URI;

@Service
public class CustomerRegistrationServiceImpl implements CustomerRegistrationService {

    private final CustomerCrudRepository repository;
    private final PersonRegistrationService personRegistrationService;

    public CustomerRegistrationServiceImpl(CustomerCrudRepository repository,
                                           PersonRegistrationService personRegistrationService
    ) {
        this.repository = repository;
        this.personRegistrationService = personRegistrationService;
    }

    @Override
    public ResponseEntity<CustomerResponseDTO> register(@Valid PersonRegistrationDTO registrationDTO) {
        validate(registrationDTO);
        Person savedPerson = personRegistrationService.register(registrationDTO);
        Customer newCustomer = Customer.of(savedPerson);
        Customer savedCustomer = repository.save(newCustomer);
        CustomerResponseDTO responseDto = savedCustomer.toDto();
        return ResponseEntity
                .created(URI.create("/api/v1/customers"))
                .contentType(MediaType.APPLICATION_JSON)
                .body(responseDto);
    }

    private void validate(PersonRegistrationDTO registrationDto) {
        if (repository.findBySIN(registrationDto.sin()).isPresent()) {
            throw new DuplicatedSINException();
        }
        if (repository.findByEmail(registrationDto.email()).isPresent()) {
            throw new DuplicatedEmailException();
        }
    }
}

File: v2/src/main/kotlin/com/api/v2/customers/services/CustomerRegistrationServiceImpl.kt
package com.api.v2.customers.services

import com.api.v2.customers.domain.exposed.Customer
import com.api.v2.customers.domain.CustomerCrudRepository
import com.api.v2.customers.responses.CustomerResponseDTO
import com.api.v2.people.exceptions.DuplicatedSINException
import com.api.v2.people.requests.PersonRegistrationDTO
import com.api.v2.people.services.exposed.PersonRegistrationService
import com.api.v2.customers.utils.exposed.toDTO
import jakarta.validation.Valid
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import java.net.URI

@Service
class CustomerRegistrationServiceImpl(
    private val repository: CustomerCrudRepository,
    private val personRegistrationService: PersonRegistrationService
) : CustomerRegistrationService {

    override fun register(registrationDTO: @Valid PersonRegistrationDTO): ResponseEntity<CustomerResponseDTO> {
        validate(registrationDTO)
        val savedPerson = personRegistrationService.register(registrationDTO)
        val newCustomer = Customer.update(savedPerson)
        val savedCustomer = repository.save<Customer>(newCustomer)
        val dto = savedCustomer.toDTO()
        return ResponseEntity
            .created(URI.create("/api/v2/customers"))
            .contentType(MediaType.APPLICATION_JSON)
            .body(dto)
    }

    private fun validate(registrationDTO: PersonRegistrationDTO) {
        if (repository.findBySIN(registrationDTO.sin) != null) {
            throw DuplicatedSINException()
        }
        if (repository.findByEmail(registrationDTO.email) != null) {
            throw DuplicatedSINException()
        }
    }
}